src/resolve/resolved-dns-server.c

```diff
@@ -518,11 +520,42 @@ int dns_server_adjust_opt(DnsServer *server, DnsPacket *packet, DnsServerFeature
         return dns_packet_append_opt(packet, packet_size, edns_do, NULL);
 }
 
+int dns_server_ifindex(const DnsServer *s) {
+        assert(s);
+
+        /* The link ifindex always takes precedence */
+        if (s->link)
+                return s->link->ifindex;
+
+        if (s->ifindex > 0)
+                return s->ifindex;
+
+        return 0;
+}
+
 const char *dns_server_string(DnsServer *server) {
         assert(server);
 
-        if (!server->server_string)
-                (void) in_addr_to_string(server->family, &server->address, &server->server_string);
+        if (server->server_string)
+                return server->server_string;
+
+        (void) in_addr_to_string(server->family, &server->address, &server->server_string);
+
+        if (server->server_string &&
+            server->family == AF_INET6 &&
+            in_addr_is_link_local(server->family, &server->address) &&
+            dns_server_ifindex(server) > 0) {
+
+                char *e;
+
+                /* Optionally append the interface index, if this is IPv6 and the address is link-local and hence
+                 * requires the interface index to be useful */
+
+                if (asprintf(&e, "%s%%%i", server->server_string, dns_server_ifindex(server)) >= 0) {
+                        free(server->server_string);
+                        server->server_string = e;
```

Comment:
I think this is bad style. server->server_string could end up with partial information if the error happens. In practice this is very unlikely, but it's just ugly.

Why not add `in_addr_iface_to_string()` instead, which does the conversion atomically? It'll probably find other uses too.